File: crabzilla-core/src/main/java/io/github/crabzilla/core/entity/EntityCommandResult.java
package io.github.crabzilla.core.entity;

import lombok.EqualsAndHashCode;
import lombok.NonNull;

import java.util.function.Consumer;

@EqualsAndHashCode
public class EntityCommandResult {

  private final EntityUnitOfWork unitOfWork;
  private final Throwable exception;

  private EntityCommandResult(EntityUnitOfWork unitOfWork, Throwable exception) {
    this.unitOfWork = unitOfWork;
    this.exception = exception;
  }

  public void inCaseOfSuccess(@NonNull Consumer<EntityUnitOfWork> consumer) {
    if (unitOfWork != null) {
      consumer.accept(unitOfWork);
    }
  }

  public void inCaseOfError(@NonNull Consumer<Throwable> consumer) {
    if (exception != null) {
      consumer.accept(exception);
    }
  }

  static public EntityCommandResult success(@NonNull EntityUnitOfWork uow) {
    return new EntityCommandResult(uow, null);
  }

  static public EntityCommandResult error(@NonNull Throwable e) {
    return new EntityCommandResult(null, e);
  }

}

File: crabzilla-core/src/main/java/io/github/crabzilla/core/entity/EntityCommandResult.kt
package io.github.crabzilla.core.entity

class EntityCommandResult private constructor(private val unitOfWork: EntityUnitOfWork?,
                                              private val exception: Throwable?) {

  fun inCaseOfSuccess(uowFn: (EntityUnitOfWork?) -> Unit) {
    if (unitOfWork != null) {
      uowFn.invoke(unitOfWork)
    }
  }

  fun inCaseOfError(uowFn: (Throwable) -> Unit) {
    if (exception != null) {
      uowFn.invoke(exception)
    }
  }

  companion object {

    fun success(uow: EntityUnitOfWork): EntityCommandResult {
      return EntityCommandResult(uow, null)
    }

    fun error(e: Throwable): EntityCommandResult {
      return EntityCommandResult(null, e)
    }
  }

}
